server: Fix NPE in spanStatsFanOut

kyle-a-wong opened the pull request:
a NPE was surfaced when doing a spanstats request on an unfinalized (mixed version) cluster.

To fix, defensive checks are added to defend against potential nil responses and references to non-existant map entries for a request span stat

Fixes: #132130
Release note (bug fix): Fixes a bug where a span stats request on a mixed version cluster resulted in an NPE

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/132349)
<!-- Reviewable:end -->

dhartunian reviewed (commented):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132349)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @angles-n-daemons, @kyle-a-wong, and @xinhaoz)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132349#-O8wChjkCIwT_I9oe6r0):*
nit: this wrap feels too narrow
___
*[`pkg/server/span_stats_server.go` line 146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132349#-O8wCqJZ7v31b5-EsF1_:-O8wCqJZ7v31b5-EsF1a:bhhrgrs) ([raw file](https://github.com/cockroachdb/cockroach/blob/d36dbc88da910972c6ee5043e9cb75d8cacd95d7/pkg/server/span_stats_server.go#L146)):*
> ```Go
> 		// See #108779.
> 		for spanStr, spanStats := range nodeResponse.SpanToStats {
> 			if spanStats == nil {
> ```

can you pull out the loop body into a function so it can be unit tested? I see how testing the whole handler with this change is cumbersome, but it would be nice to have some simple coverage to validate the fix.


<!-- Sent from Reviewable.io -->

kyle-a-wong committed:
server: Fix NPE in spanStatsFanOut

a NPE was surfaced when doing a span stats request on an unfinalized
(mixed version) cluster.

To fix, defensive checks are added to defend against potential nil
responses and references to non-existant map entries for a request
span stat

Fixes: #132130
Release note (bug fix): Fixes a bug where a span stats request on a
mixed version cluster resulted in an NPE

kyle-a-wong reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132349)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @angles-n-daemons, @dhartunian, and @xinhaoz)

___
*[`pkg/server/span_stats_server.go` line 146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132349#-O8wCqJZ7v31b5-EsF1_:-O8xCcEI-8F62jXXhKiV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d36dbc88da910972c6ee5043e9cb75d8cacd95d7/pkg/server/span_stats_server.go#L146)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

can you pull out the loop body into a function so it can be unit tested? I see how testing the whole handler with this change is cumbersome, but it would be nice to have some simple coverage to validate the fix.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I think you pulled out a wider scope than I initially requested, but I see how that was perhaps easier to do for testing.

> Reviewed 1 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132349#-:-O9FTOBmDlRZWFoRCSEG:b-w1yx5o)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @angles-n-daemons and @xinhaoz)



<!-- Sent from Reviewable.io -->

kyle-a-wong commented:
tftr!

bors r+
